react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

```diff
@@ -0,0 +1,384 @@
+
+Another question that might pop up in your mind is, "Why not just use `querySelector` or other DOM manipulation methods that we've done previously in vanilla JavaScript?" Primarily because of performance and using the hook with React, it's ensured that it will be optimized. We should also not stray outside React and use the provided API to work on their ecosystem.
+
+We can also see that it's similar to the `useState` hook in that it can store some values. The main difference is that `useRef` creates a mutable reference, allowing you to update its value without triggering a re-render. But, `useState` manages an immutable state that triggers re-renders when updated.
+
+### The useMemo hook
+
+In all of the examples, I would advise you to use the [Profiler component](https://react.dev/reference/react/Profiler) that is provided in the `react` module. If you want a more interactive alternative, use the `Profiler` in the [React Developer Tools](https://react.dev/learn/react-developer-tools). To measure rendering performance. Note that sometimes you don't need to optimize anything because of how fast things are already. As the famous saying goes in software development:
+
+> Premature optimization is the root of all evil -- The Art of Computer Programming by Donald Knuth
+
+The `useMemo` hook provides a way to add memoization inside our components. It's used to optimize expensive or complex calculations where it caches the result of a function call and stores it to be used later without recalculating it. The memoized value is, however, recalculated *only* when the dependencies of the `useMemo` hook change. And yes, this hook's parameters are the same as the `useEffect` hook you already know. The hook takes in two arguments: a `calculateValue` callback and a `dependencies` array.
+
+#### Memoizing expensive calculations
+
+In our previous projects, namely the Shopping Cart Project. You have some logic where you calculate the total price of the products added to the cart. You might or might not have a `Cart` component that functions as a drawer, where the user can open the cart every time they either click on the `Add to Cart` button or the Cart icon in the header.
+
+An example of a `Cart` component:
+
+~~~jsx
+function Cart({ products }) {
+  const totalPrice = products.reduce(
+    (total, product) => total + product.price * product.quantity,
+    0
+  );
+
+  return (
+    <div>
+      {/* Some other content in the cart */}
+      {/* Products to display */}
+      <p>
+        Total Price: <strong>${totalPrice}</strong>
+      </p>
+      {/* Some button to checkout */}
+    </div>
+  );
+}
+~~~
+
+In our `Cart` component, we have the total price of the products calculated directly inside the component. Every time the component is rendered or updated, the calculation is performed from scratch! That doesn't sound good... What if the user has added many products to the cart? Then it will lead to a sluggish user experience.
```

Comment:
We could be more specific than "many". Maybe "hundreds of thousands" or "millions", since at that number will `useMemo` starts to become evidently useful. What do you think?

```suggestion
In our `Cart` component, we have the total price of the products calculated directly inside the component. Every time the component is rendered or updated, the calculation is performed from scratch! That doesn't sound good... What if the user has added hundreds of thousands of products to the cart? Then it will lead to a sluggish user experience.
```

Reply:
Sounds good, I wrote this thinking of letting the reader interpret "many", but yeah could be more specific 👍